src/utils/codegen.cr

```diff
@@ -0,0 +1,235 @@
+module Mint
+  module Codegen
+    alias Node = String |
+                 SourceMappable |
+                 NodeContainer |
+                 NodeIndent |
+                 NodeNoIndent |
+                 NodeStrip
+
+    class NodeContainer
+      def initialize(@nodes : Array(Node))
+      end
+
+      def to_s(io)
+        io << (Codegen.build self)[:code]
+        raise "Please use Codegen.join to concatenate Strings with nodes."
+      end
+    end
+
+    class SourceMappable
+      def initialize(@ast_node_from : Ast::Node?, @ast_node_to : Ast::Node, @node : Node, @is_symbol : Bool)
+        symbol.try do |s|
+          if s.includes?('\n')
+            raise "symbol is supposed to be just a name, e.g. for a variable or a function but it is: #{s}"
+          end
+        end
+      end
+
+      def symbol
+        @is_symbol ? file.input[from, to - from] : nil
+      end
+
+      def from
+        (@ast_node_from || @ast_node_to).from
+      end
+
+      def to
+        @ast_node_to.to
+      end
+
+      def file
+        @ast_node_to.input
+      end
+
+      def to_s(io)
+        raise "Please use Codegen.join to concatenate Strings with nodes."
+      end
+    end
+
+    class NodeIndent
+      def initialize(@child : Node, @spaces : Int32)
+      end
+    end
+
+    class NodeNoIndent
+      def initialize(@child : Node)
+      end
+    end
+
+    class NodeStrip
+      def initialize(@child : Node)
+      end
+    end
+
+    def self.empty?(node : Node)
+      case node
+      in NodeContainer
+        node.@nodes.all? { |n| empty?(n) }
```

Comment:
Please avoid accessing @ivars from outside of the object, use getters for such properties.